src/Components/Server/src/Circuits/RemoteJSDataStream.cs

```diff
@@ -0,0 +1,173 @@
+
+using System;
+using System.Buffers;
+using System.IO;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.JSInterop;
+
+namespace Microsoft.AspNetCore.Components.Server.Circuits
+{
+    internal sealed class RemoteJSDataStream : Stream
+    {
+        private readonly RemoteJSRuntime _runtime;
+        private readonly long _streamId;
+        private readonly long _totalLength;
+        private readonly CancellationToken _streamCancellationToken;
+        private readonly Stream _pipeReaderStream;
+        private readonly Pipe _pipe;
+        private long _bytesRead;
+
+        public static async Task<bool> ReceiveData(RemoteJSRuntime runtime, long streamId, byte[] chunk, string error)
+        {
+            if (!runtime.RemoteJSDataStreamInstances.TryGetValue(streamId, out var instance))
+            {
+                // There is no data stream with the given identifier. It may have already been disposed.
+                // We notify JS that the stream has been cancelled/disposed.
+                return false;
+            }
+
+            return await instance.ReceiveData(chunk, error);
+        }
+
+        public static async ValueTask<RemoteJSDataStream> CreateRemoteJSDataStreamAsync(
+            RemoteJSRuntime runtime,
+            IJSDataReference jsDataReference,
+            long totalLength,
+            long maxBufferSize,
+            long maximumIncomingBytes,
+            CancellationToken cancellationToken = default)
+        {
+            var streamId = runtime.RemoteJSDataStreamNextInstanceId++;
+            var remoteJSDataStream = new RemoteJSDataStream(runtime, streamId, totalLength, maxBufferSize, cancellationToken);
+            await runtime.InvokeVoidAsync("Blazor._internal.sendJSDataStream", jsDataReference, streamId, maximumIncomingBytes);
+            return remoteJSDataStream;
+        }
+
+        private RemoteJSDataStream(
+            RemoteJSRuntime runtime,
+            long streamId,
+            long totalLength,
+            long maxBufferSize,
+            CancellationToken cancellationToken)
+        {
+            _runtime = runtime;
+            _streamId = streamId;
+            _totalLength = totalLength;
+            _streamCancellationToken = cancellationToken;
+
+            _runtime.RemoteJSDataStreamInstances.Add(_streamId, this);
+
+            _pipe = new Pipe(new PipeOptions(pauseWriterThreshold: maxBufferSize, resumeWriterThreshold: maxBufferSize / 2));
+            _pipeReaderStream = _pipe.Reader.AsStream();
+        }
+
+        // Ideally we'd have IAsyncEnumerable<ReadOnlySequence<byte>> here so we can pass through the
+        // data without having to copy it into a temporary buffer in BlazorPackHubProtocolWorker. But trying
+        // this gives strange errors; sometimes the "chunk" variable below has a negative length, even though
+        // the logic never returns a corrupted item as far as I can tell.
```

Comment:
```suggestion
        // Ideally we'd have IAsyncEnumerable<ReadOnlySequence<byte>> here so we can pass through the
        // data without having to copy it into a temporary buffer in BlazorPackHubProtocolWorker. But trying
        // this gives strange errors; sometimes the "chunk" variable below has a negative length, even though
        // the logic never returns a corrupted item as far as I can tell.
```

I think we can drop this comment if we're using `byte[]` and don't specifically have a plan to change to `IAsyncEnumerable<ReadOnlySequence<byte>>`.